source-mongodb/pull.go

```diff
@@ -0,0 +1,267 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+	"math"
+	"errors"
+	"io"
+
+	boilerplate "github.com/estuary/connectors/source-boilerplate"
+	pc "github.com/estuary/flow/go/protocols/capture"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	"golang.org/x/sync/errgroup"
+	log "github.com/sirupsen/logrus"
+
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Pull is a very long lived RPC through which the Flow runtime and a
+// Driver cooperatively execute an unbounded number of transactions.
+func (d *driver) Pull(stream pc.Driver_PullServer) error {
+	var open, err = stream.Recv()
+	if err != nil {
+		return fmt.Errorf("error reading PullRequest: %w", err)
+	} else if open.Open == nil {
+		return fmt.Errorf("expected PullRequest.Open, got %#v", open)
+	}
+
+	var cfg config
+	if err := pf.UnmarshalStrict(open.Open.Capture.EndpointSpecJson, &cfg); err != nil {
+		return fmt.Errorf("parsing config json: %w", err)
+	}
+
+	var prevState captureState
+	if open.Open.DriverCheckpointJson != nil {
+		if err := pf.UnmarshalStrict(open.Open.DriverCheckpointJson, &prevState); err != nil {
+			return fmt.Errorf("parsing state checkpoint: %w", err)
+		}
+	}
+
+	var ctx = context.Background()
+
+	client, err := d.Connect(ctx, cfg)
+	if err != nil {
+		return fmt.Errorf("connecting to database: %w", err)
+	}
+	log.Info("connected to database")
+
+	defer func() {
+		if err = client.Disconnect(ctx); err != nil {
+			panic(err)
+		}
+	}()
+
+	var c = capture{
+		Output: &boilerplate.PullOutput{Stream: stream},
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	if err := c.Output.Ready(); err != nil {
+		return err
+	}
+
+	for idx, binding := range open.Open.Capture.Bindings {
+		var res resource
+		if err := pf.UnmarshalStrict(binding.ResourceSpecJson, &res); err != nil {
+			return fmt.Errorf("parsing resource config: %w", err)
+		}
+		var resState, _ = prevState.Resources[resourceId(res)]
+
+		// Only backfill the first time, and we know that when there is no state
+		if len(prevState.Resources) == 0 {
+			if err := c.BackfillCollection(ctx, client, uint32(idx), res); err != nil {
+				return fmt.Errorf("backfill: %w", err)
+			}
+		}
+
+		eg.Go(func() error { return c.ChangeStream(ctx, client, uint32(idx), res, resState) })
+	}
+
+	if err := eg.Wait(); err != nil && !errors.Is(err, io.EOF) {
+		return err
+	}
+
+	return nil
+}
+
+type capture struct{
+	Output   *boilerplate.PullOutput
+	BackfillFinishedAt *primitive.Timestamp
```

Comment:
This feels like the wrong way to plumb the "when did the backfill complete" information from `BackfillCollection` to `ChangeStream` -- when there are multiple collection bindings I think there's a race condition if backfills complete quickly and/or the change streaming goroutines take longer than expected to get scheduled?

Perhaps this should simply be a return value from `c.BackfillCollection()` which is then provided as an argument to `c.ChangeStream()`?